org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/util/nodeselection/SelectionHistoryManager.java

```diff
@@ -40,32 +51,61 @@ private void init() {
         this.treeViewer.addSelectionChangedListener(this.listener);
     }
 
+    public void dispose() {
+        this.treeViewer.removeSelectionChangedListener(this.listener);
+    }
+
+    @Override
+    public void addSelectionChangedListener(ISelectionChangedListener listener) {
+        this.selectionChangedListeners.add(listener);
+    }
+
+    @Override
+    public void removeSelectionChangedListener(ISelectionChangedListener listener) {
+        this.selectionChangedListeners.remove(listener);
+    }
+
     public NodeSelection getSelectionHistory() {
         return this.selectionHistory;
     }
 
-    private void handleSelection(IStructuredSelection selection) {
+    @Override
+    public ISelection getSelection() {
+        return getSelectionHistory();
+    }
+
+    @Override
+    public void setSelection(ISelection selection) {
+        handleSelection(selection);
+    }
+
+    protected void handleSelection(ISelection selection) {
         NodeSelection nodeSelection = NodeSelection.from(selection);
         this.selectionHistory = this.selectionHistory.mergeWith(nodeSelection);
+        fireSelectionChanged();
     }
 
-    public void dispose() {
-        this.treeViewer.removeSelectionChangedListener(this.listener);
+    protected void fireSelectionChanged() {
```

Comment:
Can you please elaborate on why we need a concurrent list and a SafeRunner in this case?